sycl/doc/Assert.md

```diff
@@ -0,0 +1,421 @@
+```
+@AssertHappenedMem = global AssertHappened
+
+/// user's code
+void user_func(int X) {
+if (!(X && “X is nil")) {
+    __assert_fail(...);
+  }
+}
+
+kernel(...) {
+  do_smth()
+  user_func(0);
+  do_smth_else();
+}
+
+/// __assert_fail belongs to Linux version of devicelib
+void __assert_fail(...) {
+  ...
+  __devicelib_assert_fail(...);
+}
+
+void __devicelib_assert_fail(Expr, File, Line, GlobalID, LocalID) {
+  ...
+  volatile int *Ptr = (volatile int *)AssertHappenedMem.Flag;
+  int Expected = 0;
+  int Desired = 1;
+
+  if (atomic_CAS(&AssertHappenedMem.Flag, Expected, Desired))
+    printf("Assertion `%s' failed in %s at line %i. GlobalID: %i, LocalID: %i",
+           Expr, File, Line, GlobalID, LocalID);
+}
+```
+
+#### Compiling with assert enabled/disabled
+
+Consider the following example sources:
+```c++
+// impl.cpp
+using namespace sycl;
+int calculus(int X) {
+  assert(X && "Invalid value");
+  return X * 2;
+}
+
+void enqueueKernel(queue &Q, buffer &B) {
+  Q.submit([](handler &H) {
+    auto Acc = B.get_access<read_write>(H);
+    H.parallel_for(/* range */, [](item It) {
+      assert(Acc[It]);
+      // ...
+    });
+  });
+}
+
+// main.cpp
+// ...
+using namespace sycl;
+
+SYCL_EXTERNAL int calculus(int);
+void enqueueKernel(queue&, buffer&);
+
+void workload() {
+  queue Q;
+  buffer B;
+
+  Q.submit([](handler &H) {
+    auto Acc = B.get_access<read_write>(H);
+    H.parallel_for(/* range */, [](item It) {
+      int X = calculus(0); // should fail assertion
+      assert(X && "Nil in result");
+      Acc[It] = X;
+    });
+  });
+
+  enqueueKernel(Q, B);
+  ...
+}
+```
+
+These two files are compiled into a single binary application. There are four
+states of definedness of `NDEBUG` macro available:
+
+| # | `impl.cpp` | `main.cpp` |
+| - | ---------- | ---------- |
+| 1 | defined    | defined    |
+| 2 | defined    | undefined  |
+| 3 | undefined  | defined    |
+| 4 | undefined  | undefined  |
+
+States of definedness of `NDEBUG` macro defines the set of assertions which can
+fail.
+
+### Raising assert failure flag and reading it on host
+
+Each and every translation unit provided by user should have declaration of
+assert flag read function available:
+```c++
+int __devicelib_assert_read(void);
+```
+Also, the [AssertHappened](#prog-scope-var-decl) structure type should be
+available for the copier kernel.
+
+The definition is only provided within devicelib along with
+`__devicelib_assert_fail` function which raises the flag.
+
+Reading of assert failure flag is performed with the help of auxiliary kernel
+which is enqueued as dependent on user's one. The flag state is checked later
+in host-task. This is achieved with approximately the following changes:
+
+```c++
+#include <assert_happened.hpp> // contains extern decl of AssertHappenedMem
+
+#ifndef NDEBUG
+class AssertFlagCopier;
+#endif
+
+class queue {
+  template <typename T> event submit(T CGF) {
+    event Event = submit_impl(CGF);
+#ifndef NDEBUG
```

Comment:
queue::submit implementation is likely already located within sycl.so or can be moved there. If so, user will have asserts supported only with non-release builds of sycl runtime, because of this `#ifndef NDEBUG`. I don't think this code should be guarded by `NDEBUG`, as the `isAssertUsed(KernelName)` check is very lightweight and won't have any noticeable performance impact.

In return, we get user asserts working with any version of sycl.so.

Reply:
Implementation of `queue::submit` is header based due it is template.
This solution is chosen as easiest one to allow for copier kernel get represented in integration header when the user compiles an application.

Reply:
My point is that we don't need NDEBUG check, as we already have the lightweight `if (!get_device()->assert_fail_supported() && isAssertUsed(KernelName))` guard. NDEBUG adds risks - we should not assume device code and host code is compiled with the same NDEBUG macro setting. Does removing NDEBUG breaks anything?

Reply:
> NDEBUG adds risks - we should not assume device code and host code is compiled with the same NDEBUG macro setting.

I think it's assumed today and this might be a big feature to ask if feasible. Support different NDEBUG values for host and device compilation might lead to ODR violation for certain implementations of C++ standard library.
According to my understanding compiler tries to keep the same value for all macro-definitions except `__SYCL_DEVICE_ONLY__` to make sure that both host and device compilers see the same code.
+@erichkeane, @AaronBallman, to comment on that.

Reply:
Consider the case where a kernel calls some device function that is defined with `SYCL_EXTERNAL` in another translation unit.  The TU launching the kernel might be compiled with `NDEBUG` while the TU defining the `SYCL_EXTERNAL` function might be compiled w/o `NDEBUG`.  In such a case, we still need to enable assertion handing in the call to `queue::submit()` despite the fact that this TU has `NDEBUG` defined.  Therefore, I think @kbobrovs's suggestion is the correct way.

Reply:
@gmlueck, I think you are describing different scenario, where two different TUs are compiled by the *device* compiler with different NDEBUG value. I agree that this should be supported. My comment was about using different NDEBUG values while compiling the same TU by host and device compilers.

Reply:
@bader, I agree that it is a different scenario, but I think it is a stronger motivation for making the change that @kbobrovs proposes.

Reply:
Differing `NDEBUG` isn't itself traditionally a ODR violation, but gets you about 98% to one.  Libraries typically have to do a TON of work to make sure they don't go the other 2% if they decide to support that mechanism.  You'll note that the MSVC runtime libraries DON'T do any particular work, and instead just have you link with the one that matches your debug-setup.  The Linux libraries have to be pretty defensive of that.

In the SYCL host/device model, we consider both compiles to be the same 'translation unit' for a variety of reasons, AND we don't defend our libraries against `NDEBUG` anyway, so I think we should just as a policy consider that a violation.

Reply:
> we should not assume device code and host code is compiled with the same NDEBUG macro setting.

That seems like an incredible amount of risk for very limited reward. The host and device code reside within the same source file, and users are used to thinking of source files as being the translation unit the compiler sees. SYCL plays a bit fast and loose with that notion, but I think the general rule should be that the host and device compilers see the same code. The alternative is a language that's really hard for users to reason about (with ODR violations likely as the result).

Reply:
Maybe this is a risk, but this is how it is likely happen in real life. Is there really a big risk for SPIRV-based devices, or is it mostly for host/CPU device which may have connection points to the host code other than kernel invocation?

In any case, my proposal is removing the `#ifndef NDEBUG` from the `submit` function code which checks for `assert` presence in the device code. I wonder what additional risks does it incur?

Reply:
Removed use of `#ifdef` here. @gmlueck 's use-case sounds quite reasonable.